File: day12/app/assembunny.py
class Assembunny(object):

  def __init__(self):
    self.a = self.b = self.d = self.pos = 0
    self.c = 1

  def execute(self, code):
    while self.pos < len(code):
      self.process(code[self.pos])

  def process(self, line):
    instruction = line[0]

    if instruction == 'cpy':
      self.copy(line[1], line[2])
    elif instruction == 'inc':
      self.inc(line[1])
    elif instruction == 'dec':
      self.dec(line[1])
    elif instruction == 'jnz':
      self.jnz(line[1], line[2])

  def copy(self, newval, register):
    if type(newval) == int:
      setattr(self, register, newval)
    else:
      setattr(self, register, getattr(self, newval))
    self.pos = self.pos + 1

  def inc(self, register):
    oldval = getattr(self, register)
    setattr(self, register, oldval + 1)
    self.pos = self.pos + 1

  def dec(self, register):
    oldval = getattr(self, register)
    setattr(self, register, oldval - 1)
    self.pos = self.pos + 1

  def jnz(self, val, jump):
    if type(val) != int: val = getattr(self, val)
    if val == 0:
      self.pos = self.pos + 1
    else:
      self.pos = self.pos + jump

File: day12_elixir/lib/Assembunny.ex
defmodule Assembunny do

  def process(registers, instruction, pos \\ 0)

  def process(registers, {:cpy, val, reg}, pos) do
    if Map.has_key?(registers, val) do
      {%{ registers | reg => registers[val] }, pos + 1}
    else
      {%{ registers | reg => val }, pos + 1}
    end
  end

  def process(registers, {:inc, reg}, pos) do
    {%{ registers | reg => registers[reg] + 1 }, pos + 1}
  end

  def process(registers, {:dec, reg}, pos) do
    {%{ registers | reg => registers[reg] - 1 }, pos + 1}
  end

  def process(registers, {:jnz, val, jump}, pos) do
    cond do
      is_integer(val) && val > 0 ->
        {registers, pos + jump}
      is_atom(val) && registers[val] > 0 ->
        {registers, pos + jump}
      true ->
        {registers, pos + 1}
    end
  end

  def execute(registers, code, pos \\ 0)

  def execute(registers, code, pos) when pos < length(code) do
    {registers, newpos} = process(registers, Enum.at(code, pos), pos)
    execute(registers, code, newpos)
  end

  def execute(registers, _, _) do
    registers
  end

end
